core/src/main/java/org/elasticsearch/index/engine/IndexSearcherWrappingService.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.search.IndexSearcher;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.index.engine.Engine.Searcher;
+
+import java.util.Set;
+
+/**
+ * Service responsible for wrapping the {@link DirectoryReader} of a {@link Searcher} with the
+ * configured {@link IndexSearcherWrapper} instances.
+ */
+public final class IndexSearcherWrappingService {
+
+    private final IndexSearcherWrapper wrapper;
+
+    // for unit tests:
+    IndexSearcherWrappingService() {
+        this.wrapper = null;
+    }
+
+    @Inject
+    // Use a Set parameter here, because constructor parameter can't be optional
+    // and I prefer to keep the `wrapper` field final.
+    public IndexSearcherWrappingService(Set<IndexSearcherWrapper> wrappers) {
+        if (wrappers.size() > 1) {
+            throw new IllegalStateException("wrapping of the index searcher by more than one wrappers is forbidden");
```

Comment:
maybe add the set of wrappers to the message so taht it would be easier to debug